### ISSUE TITLE
[Feature] Support for the new Apple compact keyboard [sf#191]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#215]
**Reported by sd-snatcher on 2009-10-28 23:27 UTC**
Please add support for the new Apple compact keyboard. This keyboard comes by default with all new iMacs.
http://store.apple.com/us/product/MB869LL/A

The Apple Wireless keyboard also features this same layout.

1) Function keys don't work on windowed mode, only on fullscreen. (*1)
2) The compact keyboard don't have many legacy keys, like HOME, END, PGUP, PGDOWN. Those functions are now made by key combinations:

PGUP/PGDOWN = FN+UP/FN+DOWN (done by the keyboard firmware)
HOME/END = cmd+LEFT/cmd+RIGHT (done by software)

*1: This is probably affecting the non-compact new Apple keyboard too:  http://store.apple.com/us/product/MB110LL/A?fnode=MTY1NDA1Mg&mco=MTA4NzM1Nzc

**Comment by openMSX-import** [comment]
**Commented by mthuurne on 2009-10-28 23:44 UTC**
I don't have this keyboard, so I'll need very detailed information on what kind of changes are needed. Which hotkeys and/or MSX keys of openMSX do not work with this keyboard? Note that we already have different a hotkey mapping under OS X because F9 and up are claimed by the OS. For example throttle is F9 under Windows and X11 but Cmd-T under OS X. Do F1-F5 work as MSX keys if you press Fn + F1-F5?

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2009-10-31 21:15 UTC**
Ok, I did more extensive tests on the keyboard as you asked:

1) F1-F5 do work as expected when pressed with the "fn" key. Quite annoying to use it like this, but it works.
2) SELECT works well as fn+F7
3) STOP works partially as fn+F8: In windowed mode, when you try to use fn+CTRL+F8 to issue a CTRL+STOP to a BASIC program, the menu of the status-bar opens.

I disabled the CTRL+F8 shortcut on the Control Panel as a work-around.

I did found a very nice application, called FunctionFlip, that allowed me to configure each Fx key to work as regular Fx keys, inverting the use of the fn key. If anybody needs this app, it can be found here:
http://kevingessner.com/software/functionflip/

I believe this settles the keyboard problem.   :)
